File: Leetcode 3501-3600/3574/3574.py
# Leetcode 3574: Maximize Subarray GCD Score
# https://leetcode.com/problems/maximize-subarray-gcd-score/
# Solved on 6th of December, 2025
import math


class Solution:
    def maxGCDScore(self, nums: list[int], k: int) -> int:
        """
        Calculates the maximum possible GCD score of a subarray.

        Args:
            nums: A list of integers.
            k: An integer representing the maximum allowed cost.

        Returns:
            The maximum possible GCD score.
        """
        n = len(nums)
        maxScore = 0

        for i in range(n):
            if (n - i) * nums[i] * 2 <= maxScore:
                continue

            currG = nums[i]
            cost = 1

            currScore = currG
            if cost <= k:
                currScore *= 2
            if currScore > maxScore:
                maxScore = currScore

            for j in range(i + 1, n):
                val = nums[j]

                if val % currG != 0:
                    newG = math.gcd(currG, val)
                    ratio = currG // newG
                    if ratio % 2 == 0:
                        cost = 0
                    currG = newG

                if (n - i) * currG * 2 <= maxScore:
                    break

                if (val // currG) % 2 == 1:
                    cost += 1

                currLen = j - i + 1
                score = currLen * currG
                if cost <= k:
                    score *= 2

                if score > maxScore:
                    maxScore = score

        return maxScore

File: Leetcode 3501-3600/3574/3574.java
// Leetcode 3574: Maximize Subarray GCD Score
// https://leetcode.com/problems/maximize-subarray-gcd-score/
// Solved on 6th of December, 2025
class Solution {
    /**
     * Calculates the maximum possible GCD score of a subarray.
     * @param nums The input array of integers.
     * @param k The maximum allowed cost for a subarray to be considered for double score.
     * @return The maximum GCD score.
     */
    public long maxGCDScore(int[] nums, int k) {
        int n = nums.length;
        long maxScore = 0;

        for (int i = 0; i < n; i++) {
            long currentGcd = nums[i];
            long currentCost = 1;
            
            if (currentGcd > maxScore) {
                maxScore = currentGcd;
            }
            if (currentCost <= k) {
                if (currentGcd * 2 > maxScore) {
                    maxScore = currentGcd * 2;
                }
            }

            for (int j = i + 1; j < n; j++) {
                long newGcd = gcd(currentGcd, nums[j]);
                
                if (newGcd != currentGcd) {
                    long ratio = currentGcd / newGcd;
                    if (ratio % 2 == 0) {
                        currentCost = 0;
                    }
                    currentGcd = newGcd;
                }
                
                if ((nums[j] / currentGcd) % 2 != 0) {
                    currentCost++;
                }
                
                long currentLen = j - i + 1;
                long score = currentLen * currentGcd;
                if (score > maxScore) {
                    maxScore = score;
                }
                
                if (currentCost <= k) {
                    if (score * 2 > maxScore) {
                        maxScore = score * 2;
                    }
                }
            }
        }
        return maxScore;
    }

    private long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}